libcaf_core/caf/blocking_fan_out_response_handle.hpp

```diff
@@ -0,0 +1,276 @@
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/main/LICENSE.
+
+#pragma once
+
+#include "caf/abstract_blocking_actor.hpp"
+#include "caf/actor_clock.hpp"
+#include "caf/detail/expected_builder.hpp"
+#include "caf/detail/response_type_check.hpp"
+#include "caf/message_id.hpp"
+#include "caf/policy/select_all.hpp"
+#include "caf/policy/select_any.hpp"
+#include "caf/sec.hpp"
+
+#include <type_traits>
+
+namespace caf::detail {
+
+template <class Policy, class Result>
+struct blocking_fan_out_response_handle_oracle;
+
+template <class Policy>
+struct blocking_fan_out_response_handle_oracle<Policy, message> {
+  using type = blocking_fan_out_response_handle<Policy, message>;
+};
+
+template <class Policy>
+struct blocking_fan_out_response_handle_oracle<Policy, type_list<void>> {
+  using type = blocking_fan_out_response_handle<Policy>;
+};
+
+template <class Policy, class... Results>
+struct blocking_fan_out_response_handle_oracle<Policy, type_list<Results...>> {
+  using type = blocking_fan_out_response_handle<Policy, Results...>;
+};
+
+template <class Policy, class Result>
+using blocking_fan_out_response_handle_t =
+  typename blocking_fan_out_response_handle_oracle<Policy, Result>::type;
+
+template <class Policy, class Result>
+struct blocking_fan_out_delayed_response_handle_oracle;
+
+template <class Policy>
+struct blocking_fan_out_delayed_response_handle_oracle<Policy, message> {
+  using type = blocking_fan_out_delayed_response_handle<Policy, message>;
+};
+
+template <class Policy>
+struct blocking_fan_out_delayed_response_handle_oracle<Policy,
+                                                       type_list<void>> {
+  using type = blocking_fan_out_delayed_response_handle<Policy>;
+};
+
+template <class Policy, class... Results>
+struct blocking_fan_out_delayed_response_handle_oracle<Policy,
+                                                       type_list<Results...>> {
+  using type = blocking_fan_out_delayed_response_handle<Policy, Results...>;
+};
+
+template <class Policy, class Result>
+using blocking_fan_out_delayed_response_handle_t =
+  typename blocking_fan_out_delayed_response_handle_oracle<Policy,
+                                                           Result>::type;
+
+} // namespace caf::detail
+
+namespace caf {
+
+/// Holds state for a blocking response handles.
+struct blocking_fan_out_response_handle_state {
+  /// Points to the parent actor.
+  abstract_blocking_actor* self;
+
+  /// Stores the ID of the message we are waiting for.
+  std::vector<message_id> mids;
+
+  /// Timeout for the response.
+  disposable in_flight;
+
+  /// Deadline for the response.
+  actor_clock::time_point deadline;
+};
+
+/// This helper class identifies an expected response message and enables
+/// `request(...).receive(...)`.
+template <class Policy, class... Results>
+class blocking_fan_out_response_handle {
+public:
```

Comment:
```C++
  static constexpr bool is_select_all = ...;

  static constexpr bool is_select_any = ...;

  static_assert(is_select_all || is_select_any);
```

Then use `requires` to dispatch instead of using `if constexpr` inside the function. Makes it IMHO easier to follow what's going than two big if-else branches.